### ISSUE TITLE
need 'const' string params

### ISSUE THREAD

**Issue by Alexey-T** [MWASoftware/oauth2client#6]
```
  TOAuth2OnAccessToken = procedure(Sender: TObject; AccessToken, RefreshToken, TokenScope: string;
    expires_in: integer) of object;
    procedure InternalGrantAuthorizationCode(Scope: string);    
    procedure InternalImplicitGrant(Scope: string);    
    function Post(URL: string; Params: TStrings; Response: TOAuth2Response): integer;    
```
you pass string-params BY VALUE (no word 'const'). it is much slower. must use `function F(const s: string)`.

**Comment by MWASoftware** [comment]
Alexey,

I have been pondering as to how to respond to this. It is a matter of 
programming style and I vary rarely use "const" function parameters in 
Pascal compared with C++. If only because they aren't really const in 
the C++ meaning. They just suspend certain run time checks and can 
result in hard to locate bugs.

In this specific case, if an application uses OAuth2 functions more than 
once each time it is run, then you have to ask: what is it doing? Minor 
performance gains are unlikely to have much of an impact. I much prefer 
the code to be "safe" rather than "just a little bit faster". If you are 
in a tight loop - the yes, this may be worth it - but not here.

Reference counted variables, such as strings: all you are really talking 
about is saving a reference count inc on entry and an internal 
try..finally and reference count dec on exit. I cannot believe that that 
will make any measurable performance difference at all, except in the 
most extreme cases of tight loops - which just is not the case here.

Also, you have highlighted a callback. My personal view is that const 
should never be used here. You are constraining the programmer that 
writes the callback to treat such parameters as const as well as denying 
them the opportunity to take advantage of any checks that const will 
suspend. I do not believe that is correct behaviour for a published API.

I appreciate you wanting to give advice, and do not want to appear 
negative, but I do not want to go down the path you are suggesting.

Regards

Tony

**Comment by Alexey-T** [comment]
TLDR

the “const” will not make functions less safe. But will make them faster. A little.
